Extension/src/LanguageServer/settings.ts

```diff
@@ -312,36 +309,134 @@ export class CppSettings extends Settings {
+    // Additionally, it checks for valid enum values if an array of enums is detected.
+    private getAsArrayOfStrings(settingName: string, allowUndefinedEnums: boolean = false): string[] {
+        const value: any = super.Section.get(settingName);
+        const setting = getRawSetting("C_Cpp." + settingName);
+        if (isArrayOfString(value)) {
+            if (!allowUndefinedEnums) {
+                if (setting.items.enum !== undefined) {
+                    if (!value.every(x => this.validateEnum(setting.items.enum, x))) {
+                        return setting.default;
+                    }
+                }
+            }
+            return value;
+        }
+        return setting.default;
+    }
+
+    // This helper function returns the value of a setting as a key-value object with proper type validation.
+    private getAsKeyValueObject(settingName: string, keyType: string, valueType: string): any {
+        const value: any = super.Section.get(settingName);
+        if (this.isValidMapping(value, keyType, valueType)) {
+            return value;
+        }
+        const setting = getRawSetting("C_Cpp." + settingName);
+        return setting.default;
+    }
+
+    // This helper function validates whether the given object is a valid mapping of key and value type.
+    // EX: {"key": true, "key2": false} should return true for keyType = string and valueType = boolean.
+    private isValidMapping(value: any, keyType: string, valueType: string): boolean {
+        if (value === undefined || value === null) {
+            return false;
+        }
+        return Object.entries(value).every(([key, val]) => typeof key === keyType && typeof val === valueType);
+    }
+
+    // This helper function checks a given enum value against a list of valid enum values from package.json.
+    private validateEnum(enumDescription: any, value: any): boolean {
+        if (isArray(enumDescription) && enumDescription.length > 0) {
+            return enumDescription.some(x => x.toLowerCase() === value.toLowerCase());
+        }
+        return false;
+    }
+
+    // This helper function validates whether the given value is a valid enum value for the given setting.
+    private isValidEnum(settingName: string, value: any): boolean {
+        const setting = getRawSetting("C_Cpp." + settingName);
+        if (this.validateEnum(setting.enum, value)) {
+            return value;
+        }
+        return setting.default;
+    }
+
+    private isArrayOfCommentContinuationPatterns(x: any): x is (string | CommentPattern)[] {
+        return isArray(x) && x.every(y => isString(y) || this.isCommentPattern(y));
+    }
+
+    private isCommentPattern(x: any): x is CommentPattern {
+        return isString(x.begin) && isString(x.continue);
+    }
+
+    public get maxConcurrentThreads(): number | undefined | null { return this.getAsNumber("maxConcurrentThreads"); }
```

Comment:
Since we should be ensuring a valid default comes out of our helper functions, can we remove the `| undefined | null` from most of these type declarations?

Reply:
From what I recall, we currently have special behavior surrounding null values in our native implementation. As such, I think we should not remove the usage of null values in settings for this PR iteration. What do you think?

Reply:
I'm only asking you to remove it from the ones that don't actually need it (as defined in package.json). Are you saying that all of these settings need it?

Reply:
And `undefined` should no longer be a possibility because we always return a default value. For that cases where that default value is `null`, we need to keep the `null` type annotation.